front-end/src/visualization/Visualization.jsx

```diff
@@ -0,0 +1,130 @@
+import { Box, Flex, Button } from '@chakra-ui/react';
+import Graph from './Graph';
+import Navbar from '../Navbar';
+import { useParams } from 'react-router-dom';
+import { useState } from 'react';
+
+function Visualization() {
+    const params = useParams();
+    const username = params.username;
+    const [zoomLevel, setZoomLevel] = useState(1);
+
+    const zoomIn = () => {
+        setZoomLevel(prevZoom => Math.min(prevZoom * 1.1, 2));
+    };
+    const zoomOut = () => {
+        setZoomLevel(prevZoom => Math.max(prevZoom / 1.1, 0.5));
+    };
+    const resetZoom = () => {
+        setZoomLevel(1);
+    };
+
+    function getRandomColor() {
+        const letters = '01234567';
+        let color = '#';
+        for (let i = 0; i < 6; i++) {
+            color += letters[Math.floor(Math.random() * letters.length)];
+        }
+        return color;
+    }
+
+    function calculateClusterPositions(count, centerX, centerY, radius) {
+        const positions = [];
+        for (let i = 0; i < count; i++) {
+            const angle = (i / count) * 2 * Math.PI;
+            const x = centerX + radius * Math.cos(angle);
+            const y = centerY + radius * Math.sin(angle);
+            positions.push({ x, y });
+        }
+        return positions;
+    }
+
+    function calculateCircularPositions(centerX, centerY, radius, count) {
+        const positions = [];
+        for (let i = 0; i < count; i++) {
+            const angle = (i / count) * 2 * Math.PI;
+            const x = centerX + radius * Math.cos(angle);
+            const y = centerY + radius * Math.sin(angle);
+            positions.push({ x, y });
+        }
+        return positions;
+    }
+
+    function generateClusterData(clusterId, clusterLabel, centerX, centerY, songCount, radius) {
+        const clusterColor = getRandomColor();
+        const songPositions = calculateCircularPositions(centerX, centerY, radius, songCount);
+        const nodes = [{ id: clusterId, label: clusterLabel, x: centerX, y: centerY, isCluster: true, color: clusterColor }];
+        const edges = [];
+
+        songPositions.forEach((pos, index) => {
+            const songId = clusterId * 100 + index + 1;
+            nodes.push({ id: songId, label: `Song ${index + 1}`, x: pos.x, y: pos.y, color: clusterColor });
+            edges.push({ from: clusterId, to: songId, color: 'white' });
+        });
+
+        return { nodes, edges };
+    }
+
+    function addInterClusterEdges(nodes, edges) {
+        const song1 = nodes.find(node => node.label === 'Song 1' && node.id > 100 && node.id < 200);
+        const song3 = nodes.find(node => node.label === 'Song 3' && node.id > 200 && node.id < 300);
+        if (song1 && song3) {
+            edges.push({
+                from: song1.id,
+                to: song3.id,
+                color: 'red'
+            });
```

Comment:
This will be removed so it is calculated dynamically, right?

Reply:
> This will be removed so it is calculated dynamically, right?

Yes, this was just to see how to create clusters in the frontend. Once I complete the K-means clustering algorithm in the backend this will change!